### PR TITLE
Allow expr_dynamic_cast to be used with std_code types

### PR DESCRIPTION
**Pull request by reuk**
Provides some utility methods for the types in std_code.h, which allow them to be used with expr_dynamic_cast.

Suggest @NathanJPhillips and @smowton to review.

### CODE REVIEW

File: src/util/expr_cast.h
Comment (NathanJPhillips) [inline src/util/expr_cast.h]: Does this need to be templated? Can it just take `exprt`?

File: src/util/std_code.h
Comment (NathanJPhillips) [inline src/util/std_code.h]: space after `template`?

File: src/util/std_code.h
Comment (NathanJPhillips) [inline src/util/std_code.h]: `validate_operands(x, 1, "Decls must have one or more operands", true);`
Although I think there may be a bug in `validate_operands` that you'll need to fix too.

File: src/util/std_code.h
Comment (NathanJPhillips) [inline src/util/std_code.h]: "mustn't have" seems better English.

File: src/util/std_code.h
Comment (NathanJPhillips) [inline src/util/std_code.h]: operand, not operands

**Review by NathanJPhillips — CHANGES_REQUESTED** [review]

**Comment by reuk** [comment]
Made suggested changes

File: src/util/expr_cast.h
Comment (NathanJPhillips) [inline src/util/expr_cast.h]: template specialisation comment now redundant.

**Review by NathanJPhillips — CHANGES_REQUESTED** [review]

**Comment by smowton** [comment]
Suggest adding comments indicating for each code where you *didn't* supply a validate, what prevented you from doing so

**Comment by NathanJPhillips** [comment]
On the other hand, validating invariant while casting is fairly arbitrary - you could just as well validate on member access or at any other point. Don't feel compelled to add validation before being able to merge this useful improvement as long as there wasn't validation when casting before.

**Comment by reuk** [comment]
Fixed-up, please re-review

**Review by smowton — APPROVED** [review]
Ahh, I thought they had tricky edge cases where they were sometimes used with odd operand numbers; turns out you were just mirroring the existing validation code. Thanks for clarifying!

**Review by NathanJPhillips — APPROVED** [review]

**Comment by reuk** [comment]
@smowton @NathanJPhillips I've updated this with the feature we discussed (separate `dynamic_cast` and `checked_cast` implementations), could you review the last four commits please?

File: src/util/expr_cast.h
Comment (NathanJPhillips) [inline src/util/expr_cast.h]: English. Perhaps "so that functions only"?

File: src/util/expr_cast.h
Comment (NathanJPhillips) [inline src/util/expr_cast.h]: Generally wouldn't start typedefs with a capital T, perhaps `underlying_typet`?

File: src/util/expr_cast.h
Comment (NathanJPhillips) [inline src/util/expr_cast.h]: Why isn't this infinite recursion?

File: src/util/expr_cast.h
Comment (NathanJPhillips) [inline src/util/expr_cast.h]: Is it worth creating a typedef rather than just using it inline?

File: src/util/expr_cast.h
Comment (NathanJPhillips) [inline src/util/expr_cast.h]: Is it worth creating a typedef rather than just using it inline?

File: src/util/expr_cast.h
Comment (NathanJPhillips) [inline src/util/expr_cast.h]: This operation doesn't make sense. If you assert your check will pass then why would you force yourself to write code to handle the case that you're saying can't happen?

**Review by NathanJPhillips — CHANGES_REQUESTED** [review]

**Review by NathanJPhillips — APPROVED** [review]

**Comment by NathanJPhillips** [comment]
Squash and go as far as I'm concerned.

**Pull request merged**
